vpn/client.go

```diff
@@ -0,0 +1,180 @@
+package vpn
+
+import (
+	"context"
+	"net/http"
+	"net/netip"
+	"net/url"
+
+	"golang.org/x/xerrors"
+	"nhooyr.io/websocket"
+	"tailscale.com/net/dns"
+	"tailscale.com/wgengine/router"
+
+	"github.com/tailscale/wireguard-go/tun"
+
+	"cdr.dev/slog"
+	"github.com/coder/coder/v2/codersdk"
+	"github.com/coder/coder/v2/codersdk/workspacesdk"
+	"github.com/coder/coder/v2/tailnet"
+	"github.com/coder/coder/v2/tailnet/proto"
+	"github.com/coder/quartz"
+)
+
+type Conn interface {
+	CurrentWorkspaceState() *proto.WorkspaceUpdate
+	Close() error
+}
+
+type vpnConn struct {
+	*tailnet.Conn
+
+	cancelFn    func()
+	controller  *tailnet.Controller
+	updatesCtrl *tailnet.TunnelAllWorkspaceUpdatesController
+}
+
+func (c vpnConn) CurrentWorkspaceState() *proto.WorkspaceUpdate {
+	return c.updatesCtrl.CurrentState()
+}
+
+func (c vpnConn) Close() error {
+	c.cancelFn()
+	<-c.controller.Closed()
+	return c.Conn.Close()
+}
+
+type client struct{}
+
+type Client interface {
+	NewConn(ctx context.Context, serverURL *url.URL, token string, options *Options) (Conn, error)
+}
```

Comment:
Client and the Conn are both interfaces so we can test the tunnel independently of them.